File: library0/BASIS/android/java/src/library/basis/CPtr.java
package src.library.basis;

@SuppressWarnings({W.LIB_OMIT_0, W.LIB_OMIT_1, W.LIB_OMIT_2, W.LIB_OMIT_3})
public class CPtr {

    static CPtr objectFromValue(long value) {
        CPtr ptr = new CPtr();
        ptr.mValue = value;
        return ptr;
    }

    static long valueFromObject(CPtr ptr) {
        if (ptr != null) {
            return ptr.mValue;
        } else {
            return 0;
        }
    }

    private long mValue;
}

File: library0/BASIS/android/cpp/src_library_basis/CPtr.cxx
#include "CPtr.hh"
#include "cqjnihelper.hh"

//NOTE: store c pointer with java long type.
static_assert(sizeof(jlong) >= sizeof(void *), "");

#define CLS_PATH "src/library/basis/CPtr"
CQ_JNI_CLASS(CPtr, CLS_PATH);

void *CPtr_valueFromObject(JNIEnv *env, jobject src) {
    if (env == nullptr) {
        return nullptr;
    }

    static jmethodID method = nullptr;
    cqJNIEnv::staticMethod(&method, env, CPtr(), "valueFromObject", "(L" CLS_PATH ";)J");

    return (void *)env->CallStaticLongMethod(CPtr(), method, src);
}

jobject CPtr_objectFromValue(JNIEnv *env, const void *src) {
    if (env == nullptr) {
        return nullptr;
    }

    static jmethodID method = nullptr;
    cqJNIEnv::staticMethod(&method, env, CPtr(), "objectFromValue", "(J)L" CLS_PATH ";");

    return env->CallStaticObjectMethod(CPtr(), method, (jlong)src);
}
